(DOCSP-34760): C++: Add client reset docs (#3133)
mongodb/docs-realm: source/sdk/cpp/sync/handle-sync-errors.txt

         :language: cpp

.. include:: /includes/sync-errors-in-app-services.rst

.. _cpp-client-reset:

Client Reset
------------

When using :ref:`Device Sync <sync>`, a **client reset** is an

> error recovery task that your client app must perform when given

Reviewer: "when given synced data on the server ... can no longer sync with the device" -- is it accurate that the synced _data_ is syncing with the device? or that Atlas/the server is syncing with the device?
can we revise to `when the server can longer sync...`?

synced data on the server can no longer sync with the device database. 

> In this case, the device must reset its synced database to a state that 

Reviewer: for clarity -- since we're avoiding "local", "client", "realm", etc., this gets pretty repetitious. And we can assume the device db is a synced one, no?
```suggestion
In this case, the device must reset its database to a state that 
```

matches the server in order to restore the ability to sync.

When this occurs, the unsyncable database on the device may contain data that 

> has not yet synced to the server. The SDKs can attempt to recover or 

Reviewer: ```suggestion
has not yet synced to the server. The SDK can attempt to recover or 
```
(possible discussion topic) I realize the plural `SDKs` implies that this functionality applies to `Atlas Device SDKs` as a whole, but since we're in a specific SDK's docs, should this be singular?

discard that data during the client reset process.

For more information about what might cause a client reset to occur, go to
:ref:`Client Resets in the App Services documentation <client-resets>`.

.. _cpp-automatic-vs-manual-client-reset:

Automatic vs. Manual Client Reset
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The SDKs provide :ref:`client reset modes <cpp-specify-client-reset-mode>` 

Reviewer: ```suggestion
The SDK provide :ref:`client reset modes <cpp-specify-client-reset-mode>` 
```
same

that automatically handle most client reset errors. Automatic client 
reset modes restore your device's synced database file to a syncable state 
without closing the database or missing notifications.

All the client reset modes except ``manual()`` perform an automatic client 
reset. The differences between the modes are based on how they handle 
changes on the device that have not yet synced to the backend.

Choose ``recover_unsynced_changes()`` to handle most client reset 
scenarios automatically. This attempts to recover unsynced changes when a 
client reset occurs.

In some cases, you may want or need to :ref:`set a manual client reset handler 
<cpp-manual-client-reset-handler>`. You may want to do this if your app 
requires specific client reset logic that can't be handled automatically.

.. _cpp-specify-client-reset-mode:

Specify a Client Reset Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The SDK provides the option to specify a client reset handler in your 

Reviewer: why the singular `SDK` here? if it's because this section is unique to C++, should this be `The C++ SDK`

enabled by default when you :ref:`configure Device Sync <enable-sync>`. When 
Client Recovery is enabled, the SDK can automatically manage the 
client reset process in most cases. When you make schema changes:

- The client can :ref:`recover unsynced changes <cpp-recover-unsynced-changes>` 
  when there are no schema changes, or non-breaking schema changes.
- When you make breaking schema changes, the automatic client reset modes fall
  back to a manual error handler. You can set a :ref:`manual client 
  reset error handler <cpp-manual-client-reset-handler>` for this case. 
  Automatic client recovery cannot occur when your app makes breaking 
  schema changes.

For information on breaking vs. non-breaking schema changes, see 
:ref:`breaking-change-quick-reference`.

.. _cpp-recover-unsynced-changes:

Recover Unsynced Changes
~~~~~~~~~~~~~~~~~~~~~~~~

During a client reset, client applications can attempt to recover data 
in the synced database on the device that has not yet synced to the backend. 
To recover unsynced changes, :ref:`Client Recovery <recover-unsynced-changes>` 
must be :ref:`enabled in your App Services App <enable-or-disable-recovery-mode>`, 

> as it is by default.

Reviewer: this reads awkwardly to me. Or maybe change to a parenthetical?
```suggestion
which it is by default.
```

If you want your app to recover changes that have not yet synced, use one 
of these client recovery modes:

> - ``recover_unsynced_changes()``: When you choose this mode, the client attempts 

Reviewer: ```suggestion
- ``recover_unsynced_changes()``: The client attempts 
```

  to recover unsynced changes. Choose this mode when you do not want to fall
  through to discard unsynced changes. 
- ``recover_or_discard_unsynced_changes()``: The client first attempts to 
  recover changes that have not yet synced. If the client cannot recover 
  unsynced data, it falls through to :ref:`discard unsynced changes 
  <cpp-discard-local-changes>` but continues to automatically perform the 
  client reset. Choose this mode when you want to enable automatic client 
  recovery to fall back to discard unsynced changes. 

.. literalinclude:: /examples/generated/cpp/client-reset.snippet.recover-unsynced-changes.cpp
   :language: cpp

There may be times when the client reset operation cannot complete in 

> recover unsynced changes mode, like when there are breaking schema changes or

Reviewer: might be easier to read
```suggestion
``recover_unsynced_changes()`` mode, such as when there are breaking schema changes or
```

:ref:`Client Recovery <recover-unsynced-changes>` is disabled in the 
Device Sync configuration. To handle this case, your app can handle a client
reset error in the Sync error handler. For more information, refer to 

> :ref:`manual client reset mode <cpp-manual-client-reset-handler>`.

Reviewer: ```suggestion
the :ref:`cpp-manual-client-reset-handler` section on this page.
```

.. _cpp-client-recovery-rules:

Client Recovery Rules
`````````````````````

.. include:: /includes/client-recovery-rules.rst

.. _cpp-discard-local-changes:

Discard Unsynced Changes
~~~~~~~~~~~~~~~~~~~~~~~~

> The **discard unsynced changes** client reset mode permanently deletes all 

Reviewer: [nit] none of the previous modes were formatted in bold
```suggestion
The ``discard_unsynced_changed()`` client reset mode permanently deletes all
```

unsynced changes on the device since the last successful sync. You might 
use this mode when your app requires client recovery logic that is not 

> consistent with the Device Sync :ref:`Client Recovery Rules 

Reviewer: [nit] this link text was lowercase in a previous section

<cpp-client-recovery-rules>`, or when you don't want to recover unsynced data.

Do not use discard unsynced changes mode if your application cannot lose 
device data that has not yet synced to the backend.

To perform an automatic client reset that discards unsynced changes, use the 
``discard_unsynced_changes()`` client reset mode.

.. literalinclude:: /examples/generated/cpp/client-reset.snippet.discard-unsynced-changes.cpp
   :language: cpp

.. note:: Discard with Recovery

>    If you'd like to attempt to recover unsynced changes, but but discard 

Reviewer: ```suggestion
   If you'd like to attempt to recover unsynced changes, but discard 
```

   any changes that cannot be recovered, refer to the 

>    ``recover_or_discard_unsynced_changes()`` documentation in :ref:`Recover 

Reviewer: ```suggestion
   ``recover_or_discard_unsynced_changes()`` documentation in the :ref:`cpp-recover-unsynced-changes` section on this page.
```

>    Unsynced Changes <cpp-recover-unsynced-changes>`.

Reviewer: ```suggestion
```

There may be times when the client reset operation cannot complete in 

> discard unsynced changes mode, like when there are breaking schema changes.

Reviewer: same as earlier -- might help readability 
```suggestion
``discard_unsynced_changes()`` mode, like when there are breaking schema changes.
```

To handle this case, your app can handle a client reset error in the Sync 
error handler. For more information, refer to 

> :ref:`manual client reset mode <cpp-manual-client-reset-handler>`.

Reviewer: ```suggestion
the :ref:`cpp-manual-client-reset-handler` section on this page. 
```

.. _cpp-manual-client-reset-handler:
.. _cpp-manual-mode:

Manual Client Reset Mode
~~~~~~~~~~~~~~~~~~~~~~~~

When you use ``manual()`` client reset mode, you must implement 
a custom client reset handler in the Sync error handler. We recommend using the
:ref:`automatic client recovery modes <cpp-automatic-vs-manual-client-reset>` 
when possible, and only choosing ``manual()`` mode if the automatic recovery 
logic is not suitable for your app.

.. literalinclude:: /examples/generated/cpp/client-reset.snippet.manual-client-reset.cpp
   :language: cpp

> If the client reset operation cannot complete automatically, like when there 

Reviewer: ```suggestion
If the client reset operation cannot complete automatically, such as when there 
```
